File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 140107. 점 찍기/kotlin/Solution.kt
import kotlin.math.sqrt

class Solution {
    fun solution(k: Int, d: Int): Long {
        val md2 = d.toLong() * d.toLong()
        fun d2(x: Int, y: Int): Long {
            val lx = x.toLong()
            val ly = y.toLong()
            return lx * lx + ly * ly
        }
        fun count(startEx: Int, endIn: Int) = (endIn - startEx)/k

        var answer: Long = 0
        val maxStart = sqrt(md2/2.0).toInt()
        answer += count(0, maxStart) + 1
        var d2: Long
        for (start in 0..maxStart step k) {
            var lt = start
            var rt = d
            var end = -1
            loop@while (lt <= rt) {
                val mid = (lt + rt) shr 1
                d2 = d2(mid, start)
                if (d2 == md2) {
                    end = mid
                    break@loop
                } else if (d2 < md2) {
                    lt = mid + 1
                    end = mid
                } else {
                    rt = mid - 1
                }
            }
            answer += count(start, end) shl 1
        }
        return answer
    }

}

File: 문제풀이/온라인 저지/프로그래머스/# 02. Level 2/# 140107. 점 찍기/python/solution.py
from math import sqrt
def solution(k, d):
    answer = 0

    # x 좌표를 x 라 할 때
    # y 값은 sqrt(d ** 2 - x ** 2) 보다 같거나 작아야한다. 정수 범위 내에서만 유효하므로 여기에 int 처리 (잘 모르겠으면 원 그려놓고 피타고라스 정리를 통해 구해보면 된다.)
    # 1~ max_y(위에서 구한 정수 y값) 사이에서, 가능한 점의 갯수는 이를 k 로 나눈 몫만큼이다.
    # 여기에 y = 0 일때 (x축 교차지점) 경우도 고려하여 1도 더함.
    for x in range(0, d + 1, k):
        answer += int(sqrt(d ** 2 - x ** 2))//k + 1
    return answer
